File: cpp/day6/main.cpp
#include "utils/utils.h"
#include <cassert>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <optional>
#include <sstream>
#include <tuple>
#include <vector>

enum class Direction : uint8_t {
    Up = 0x01,
    Right = 0x02,
    Down = 0x04,
    Left = 0x08
};

const uint8_t VISITED = 0x80;

Direction rotate(Direction d) {
    switch (d) {
        case Direction::Up:
            return Direction::Right;
        case Direction::Right:
            return Direction::Down;
        case Direction::Down:
            return Direction::Left;
        case Direction::Left:
            return Direction::Up;
    }

    assert(false);
}

std::optional<std::tuple<size_t, Direction>> get_next_position(std::string_view board, size_t board_width, size_t current, Direction cur_direction) {
    while (1) {
        size_t next_attempt;
        switch (cur_direction) {
            case Direction::Up:
                if (current < board_width) {
                    return std::optional<std::tuple<size_t, Direction>>();
                }
                next_attempt = current - board_width;
                break;
            
            case Direction::Right:
                next_attempt = current + 1;
                break;

            case Direction::Down:
                next_attempt = current + board_width;
                break;
            
            case Direction::Left:
                if (current == 0) {
                    return std::optional<std::tuple<size_t, Direction>>();
                }
                next_attempt = current - 1;
                break;
        }

        if (next_attempt >= board.length() || board[next_attempt] == '\n') {
            return std::optional<std::tuple<size_t, Direction>>();
        }

        switch (board[next_attempt]) {
            case '.':
            case 'X':
                return std::make_tuple(next_attempt, cur_direction);
            
            case '#':
                cur_direction = rotate(cur_direction);
                break;
            
            default:
                assert(false);
        }
    }
}

size_t trace_guard(std::string board) {
    size_t board_width = 1 + board.find('\n');
    size_t current = board.find('^');
    size_t steps = 0;
    Direction cur_direction = Direction::Up;

    board[current] = '.';

    while (1) {
        if (board[current] == '.') {
            board[current] = 'X';
            steps++;
        }

        auto result = get_next_position(board, board_width, current, cur_direction);
        if (!result) {
            break;
        }

        std::tie(current, cur_direction) = result.value();
    }

    return steps;
}

#ifdef AOC_PART2
size_t count_loops(std::string board) {
    size_t board_width = 1 + board.find('\n');
    size_t starting_position = board.find('^');
    size_t loops = 0;
    Direction starting_direction = Direction::Up;

    board[starting_position] = 'X';

    while (1) {
        auto barrier = get_next_position(board, board_width, starting_position, starting_direction);
        if (!barrier)
            break;
        
        size_t barrier_position;
        Direction next_direction;
        std::vector<uint8_t> path(board.length());
        std::tie(barrier_position, next_direction) = barrier.value();

        if (board[barrier_position] == 'X') {
            // can't put a barrier in a location we've already been in
            starting_position = barrier_position;
            starting_direction = next_direction;
            continue;
        }

        board[barrier_position] = '#';

        size_t current = starting_position;
        Direction cur_direction = starting_direction;

        while (1) {
            if ((path[current] & VISITED) && (path[current] & static_cast<uint8_t>(cur_direction))) {
                loops++;
                break;
            }

            path[current] |= VISITED;
            path[current] |= static_cast<uint8_t>(cur_direction);

            auto next = get_next_position(board, board_width, current, cur_direction);
            if (!next) {
                break;
            }

            std::tie(current, cur_direction) = next.value();
        }

        board[barrier_position] = 'X';
        starting_position = barrier_position;
        starting_direction = next_direction;
    }

    return loops;
}
#endif

int main(int argc, char *argv[]) {
    Timer t;

    if (argc < 2) {
        std::cerr << "Not enough arguments" << std::endl;
        return 1;
    }

    std::ifstream input(argv[1]);
    std::stringstream data;
    data << input.rdbuf();

#ifndef AOC_PART2
    std::cout << "Answer: " << trace_guard(data.str()) << std::endl;
#else
    std::cout << "Answer: " << count_loops(data.str()) << std::endl;
#endif
}
    

File: rs/day6/src/main.rs
use utils::timer::Timer;
use std::fs::read;
use std::ops::{BitAnd, BitOrAssign};
use std::path::PathBuf;

#[repr(u8)]
#[derive(PartialEq,Copy,Clone)]
enum Direction {
    Up = 0x08,
    Down = 0x01,
    Left = 0x02,
    Right = 0x04
}

impl Direction {
    fn rotate(&self) -> Self {
        match self {
            Self::Up => Self::Right,
            Self::Right => Self::Down,
            Self::Down => Self::Left,
            Self::Left => Self::Up
        }
    }
}

const VISITED: u8 = 0x80;

impl From<Direction> for u8 {
    fn from(value: Direction) -> Self {
        value as u8
    }
}

fn get_next_position(board: &[u8], board_width: usize, current: usize, mut cur_direction: Direction) -> Option<(usize, Direction)> {
    loop {
        let Some(next_attempt) = (match cur_direction {
            Direction::Up => current.checked_sub(board_width),
            Direction::Down => current.checked_add(board_width),
            Direction::Left => current.checked_sub(1),
            Direction::Right => current.checked_add(1),
        }) else {
            // we've underwrapped, so we've left the board
            return None;
        };

        if next_attempt >= board.len() || board[next_attempt] == b'\n' {
            // we've overshot, so we've left the board
            // (we can use the linefeed character to detect an overshot)
            return None;
        }

        match board[next_attempt] {
            b'#' => {
                cur_direction = cur_direction.rotate();
            },
            b'.' | b'X' => {
                return Some((next_attempt, cur_direction));
            },
            _ => panic!("Should not occur!")
        };
    }
}

fn trace_guard(board: &mut [u8]) -> usize {
    let board_width = 1 + board
        .iter()
        .position(|ch| *ch == b'\n')
        .expect("Invalid input");
    let starting_position = board
        .iter()
        .position(|ch| *ch == b'^')
        .expect("Missing starting position");

    // simplify logic a bit
    board[starting_position] = b'.';
    let mut cur_direction = Direction::Up;
    let mut steps = 0usize;
    let mut current = starting_position;

    loop {
        match board[current] {
            b'.' => {
                board[current] = b'X';
                steps += 1;
            },
            b'X' => {
            },
            _ => {
                panic!("Shouldn't happen!");
            }
        }

        let Some((next, next_direction)) = get_next_position(board, board_width, current, cur_direction) else {
            break;
        };

        current = next;
        cur_direction = next_direction;
    }

    steps
}

fn count_loop_points(board: &mut [u8]) -> usize {
    let board_width = 1 + board
        .iter()
        .position(|ch| *ch == b'\n')
        .expect("Invalid input");
    let mut starting_position = board
        .iter()
        .position(|ch| *ch == b'^')
        .expect("Missing starting position");

    let mut starting_direction = Direction::Up;
    let mut loops = 0usize;

    // can't put a barrier on the starting position
    board[starting_position] = b'X';

    'outer: loop {
        let mut path = vec![0u8; board.len()];
        let mut current = starting_position;
        let mut cur_direction = starting_direction;

        let Some((barrier_position, next_direction)) = get_next_position(board, board_width, current, cur_direction) else {
            // nowhere more to test
            break 'outer;
        };

        if board[barrier_position] == b'X' {
            // if we've already been here, we can't put a barrier here
            starting_position = barrier_position;
            starting_direction = next_direction;
            continue 'outer;
        }

        // simulate the barrier
        board[barrier_position] = b'#';

        'main: loop {
            if path[current].bitand(VISITED) > 0 && path[current].bitand(cur_direction as u8) > 0 {
                loops += 1;
                break 'main;
            }

            path[current].bitor_assign(VISITED);
            path[current].bitor_assign(cur_direction as u8);

            let Some((next, next_direction)) = get_next_position(board, board_width, current, cur_direction) else {
                break 'main;
            };
            
            current = next;
            cur_direction = next_direction;
        }

        board[barrier_position] = b'X';
        starting_position = barrier_position;
        starting_direction = next_direction;
    }

    loops
}

fn main() {
    let _timer = Timer::new();

    let path: PathBuf = std::env::args_os()
        .skip(1)
        .next()
        .expect("Should have file argument")
        .into();

    let mut input = read(path)
        .expect("Should be able to read from path");

    if !cfg!(feature = "part2") {
        let unique_steps = trace_guard(&mut input);
        println!("Answer: {unique_steps}");
    }
    else {
        let loop_points = count_loop_points(&mut input);
        println!("Answer: {loop_points}");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_corner_trace_guard() {
        let mut test = Vec::from(b"\
.#....
....#.
...#..
.^....
");
        assert_eq!(trace_guard(&mut test), 6);
    }

    #[test]
    fn test_corner_count_loops() {
        let mut test = Vec::from(b"\
.#....
....#.
...#..
.^....
");
        let count = count_loop_points(&mut test);
        assert_eq!(count, 1);
    }

    #[test]
    fn test_backwards_count_loops() {
        let mut test = Vec::from(b"\
..#.#.
.....#
....#.
..^...
");
        let count = count_loop_points(&mut test);
        assert_eq!(count, 1);
    }

    #[test]
    fn test_nothing() {
        let mut test = Vec::from(b"\
.###..
.#.#..
.#.#..
.#^#..
");
        let count = count_loop_points(&mut test);
        assert_eq!(count, 0);
    }

    #[test]
    fn test_no_cross_loop() {
        let mut test = Vec::from(b"\
......
...#..
....#.
......
.^#...
...#..
");
        let count = count_loop_points(&mut test);
        assert_eq!(count, 1);
    }
}
